### ISSUE TITLE
TrustSitelists flag does not work for workflow with no input

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#6999]
Andrew had issues with a TaskChain request:
https://cms-logbook.cern.ch/elog/Workflow+processing/25003

where the second task jobs were all failing with submitfailed (71101 exit code). That means there are no `possiblePSN` in the job pickle file.

The problem seems to come from the fact that the workflow was assigned with TrustSitelists=True, which simply uses the site white and black lists to figure out top level tasks location:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/JobSplitting/JobFactory.py#L178

Looking at the splitting parameters passed to the JobFactory, for the first task:

```
{'algorithm': 'EventBased', 'events_per_lumi': 1000, 'siteWhitelist': [u'T1_US_FNAL'], 'events_per_job': 1000, 'siteBlacklist': [], 'deterministicPileup': False, 'lheInputFiles': False, 'trustSitelists': True, 'performance': {'sizePerEvent': 1234.0, 'memoryRequirement': 3000.0, 'timePerEvent': 0.1}}
```

coming from the workload: `....ZTTFS_13.constraints.sites.whitelist = [u'T1_US_FNAL']`

on the other hand, the second task uses the following parameters:

```
{'algorithm': 'LumiBased', 'siteWhitelist': [], 'lumis_per_job': 10, 'deterministicPileup': False, 'siteBlacklist': [], 'trustSitelists': True, 'performance': {'sizePerEvent': 1234.0, 'memoryRequirement': 3000.0, 'timePerEvent': 0.1}}
```

coming from the workload: `...MINIAODMCUP15FS.constraints.sites.whitelist = []`

Now I see two options to fix this issue:
a) properly set the sites.whitelist attribute for all processing/production jobs (I'm afraid it can break unified job routing/overflow)
b) fail request assignment for such cases

@hufnagel @ticoann maybe you have some input on this issue?

**Comment by hufnagel** [comment]
In a TaskChain, siteWhitelist (and siteBlacklist) should only apply to the top level task. Similarly, trustSiteList should also only apply to the top level task. It has been a basic principle of our processing model that we only do site assignments for top level tasks in a workflow, anything that comes later is driven by data location.

As such, IMO an empty siteWhitelist for the second task is correct, but trustSitelists still being True is not.

**Comment by hufnagel** [comment]
Btw, trustSiteLists for the second task, is that set by us (expanding a single parameter and apply it to all tasks) or by the user ? If the former we should fix this, if the latter this is an illegal request I would say.

**Comment by hufnagel** [comment]
Also, one example that shows that your option 1 is unworkable:

Assume a TaskChain where the first task runs on opportunistic resources that stage out to another site. You absolutely need the second task to be scheduled based on data location, since you don't want to send it to the opportunistic site (it wouldn't find the data there).

**Comment by amaltaro** [comment]
> Btw, trustSiteLists for the second task, is that set by us (expanding a single parameter and apply it to all tasks) or by the user ?

The user simple sets the flag to True, so the code itself enables it for the top level task and whatever else it considers top level (further prod/proc tasks?).

If we explicitly enable it only for the first (top level task), then I see one use case where it would not work, which is when you have Task1 and Task2 processing an InputDataset, not sure we have it in real life, but it's quite possible.

**Comment by hufnagel** [comment]
Hm, can you distinguish the two use cases (second task processing the output from task 1 OR the original input again ?). Also, I would consider this illegal. Sure, you could technically do it, but why would you create a TaskChain for this ? Wouldn't it be better to make two separate workflows ?

**Comment by hufnagel** [comment]
By distinguish I mean in code. If you can, we can just put in a conditional: both siteLists and trustSiteLists apply to all tasks that read the original input to the workflow. Simple.

**Comment by amaltaro** [comment]
We used to have some HLT validation which was run with different GTs (or releases), that's one use case.
Yes, we can distinguish Task2 reading input data from "dbs/phedex" or from task1.
That's the annoying part in TaskChain, it allows you to do basically whatever you want :)

Ok, thanks Dirk. I'll have a look at this trustSitelists flag in the coming days then.

**Comment by hufnagel** [comment]
Don't just do it for trustSiteList. The siteLists themself should also apply to any other task that read the original workflow input.

**Comment by amaltaro** [comment]
And this issue strikes again in production (agent 1.3.3.patchX). Here is one example workflow:
https://cmsweb.cern.ch/reqmgr2/fetch?rid=cmsunified_task_EGM-RunIISummer19UL16GEN-00009__v1_T_200605_142821_2480

where we can see that, the workflow does not have any input dataset and still, it defines:
```
  "TrustSitelists": true, 
```

this basically forces the job factory to use the `SiteWhitelist - SiteBlacklist` as a list of `possiblePSN` for the jobs (sites where the job will get submitted to), instead of using the input file location:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/JobSplitting/JobFactory.py#L186

this information gets then consumed by JobSubmitter (and SimpleCondorPlugin), which will submit such jobs to much more sites than what it should:
https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py#L324

In short, here is what needs to be addressed in this issue:
* deny workflow assignment if `TrustSitelists=True` for a workflow without any input dataset
* if `TrustSitelists=True` for a workflow with input dataset and pileup dataset (with `TrustPUSitelists=False`), then the pileup location needs to be enforced in the list of `possiblePSN`, such that we do not submit jobs to sites that do not hold the PU data.

The pileup check will be tricky though, we do not persist pileup location anywhere besides the pileupconf.json file.
